src/main/java/org/opensearch/jobscheduler/rest/RestGetJobIndexAction.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Locale;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.opensearch.rest.RestRequest.Method.PUT;
+
+/**
+ * This class consists of the REST handler to GET job index from extensions.
+ */
+public class RestGetJobIndexAction extends BaseRestHandler {
+
+    public static final String GET_JOB_INDEX_ACTION = "get_job_index_action";
+
+    private final Logger logger = LogManager.getLogger(RestGetJobIndexAction.class);
+
+    public JobDetailsService jobDetailsService;
+
+    public RestGetJobIndexAction(final JobDetailsService jobDetailsService) {
+        this.jobDetailsService = jobDetailsService;
+    }
+
+    @Override
+    public String getName() {
+        return GET_JOB_INDEX_ACTION;
+    }
+
+    @Override
+    public List<Route> routes() {
+        return unmodifiableList(
+            asList(new Route(PUT, String.format(Locale.ROOT, "%s/%s", JobSchedulerPlugin.JS_BASE_URI, "_get/_job_index")))
+        );
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
+        XContentParser parser = restRequest.contentParser();
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+
+        GetJobIndexRequest getJobIndexRequest = GetJobIndexRequest.parse(parser);
+
+        final JobDetails[] jobDetailsResponseHolder = new JobDetails[1];
+
+        String jobIndex = getJobIndexRequest.getJobIndex();
+        String jobParameterAction = getJobIndexRequest.getJobParameterAction();
+        String jobRunnerAction = getJobIndexRequest.getJobRunnerAction();
+        String extensionId = getJobIndexRequest.getExtensionId();
+
+        CompletableFuture<JobDetails[]> inProgressFuture = new CompletableFuture<>();
+
+        jobDetailsService.processJobDetailsForExtensionId(
+            jobIndex,
+            null,
+            jobParameterAction,
+            jobRunnerAction,
+            extensionId,
+            JobDetailsService.JobDetailsRequestType.JOB_INDEX,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(JobDetails jobDetails) {
+                    jobDetailsResponseHolder[0] = jobDetails;
+                    inProgressFuture.complete(jobDetailsResponseHolder);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.info("could not process job index", e);
+                    jobDetailsResponseHolder[0] = null;
+                    inProgressFuture.completeExceptionally(e);
+                }
+            }
+        );
+
+        try {
+            inProgressFuture.orTimeout(JobDetailsService.TIME_OUT_FOR_REQUEST, TimeUnit.SECONDS);
+            inProgressFuture.get(JobDetailsService.TIME_OUT_FOR_REQUEST, TimeUnit.SECONDS);
```

Comment:
You don't need the timeout on both, they'll run sequentially.  Suggest you do
```java
inProgressFuture.orTimeout(JobDetailsService.TIME_OUT_FOR_REQUEST, TimeUnit.SECONDS).get();
```
... or `.join()` which throws different types of exceptions, doesn't throw checked exceptions, and may be preferred depending on how you choose to handle the exceptions.  Specifically, `.get()` requires you to catch `InterruptedException` and `ExecutionException`) while `.join()` does not.

Reply:
done